### PR TITLE
Branch name

### PR DESCRIPTION
**Pull request by xh3b4sd**
Fixes https://github.com/giantswarm/giantswarm/issues/1605.

### CODE REVIEW

File: cmd/deploy.go [hunk 1]
```diff
@@ -81,7 +81,9 @@ func runDeploy(cmd *cobra.Command, args []string) {
 		WorkingDirectory: workingDirectory,
 		Organisation:     organisation,
 		Project:          project,
-		Sha:              sha,
+
+		Branch: branch,
+		Sha:    sha,
```
Comment (xh3b4sd) [inline cmd/deploy.go]: Here the branch is additionally set in the project info which the sub commands receive.

File: cmd/root.go [hunk 2]
```diff
@@ -12,15 +12,17 @@ import (
 var (
 	RootCmd = &cobra.Command{
 		Use:   "architect",
-		Short: "architect is a tool for managing Giant Swarm release engineering",
+		Short: "Architect is a tool for managing builds within the Giant Swarm releases engineering.",
```
Comment (xh3b4sd) [inline cmd/root.go]: Boyscouting here. I think this is more right.

File: cmd/root.go [hunk 3]
```diff
@@ -46,17 +48,32 @@ func init() {
...
 	}
-	defaultSha := strings.TrimSpace(string(out))
 
 	RootCmd.PersistentFlags().StringVar(&workingDirectory, "working-directory", defaultWorkingDirectory, "working directory for architect")
 
 	RootCmd.PersistentFlags().StringVar(&registry, "registry", "quay.io", "docker registry")
 	RootCmd.PersistentFlags().StringVar(&organisation, "organisation", defaultOrganisation, "organisation who owns the project")
 	RootCmd.PersistentFlags().StringVar(&project, "project", defaultProject, "name of the project")
+
+	RootCmd.PersistentFlags().StringVar(&branch, "branch", defaultBranch, "git branch to build")
```
Comment (xh3b4sd) [inline cmd/root.go]: The branch is configurable via flag just like the other stuff.

File: workflow/docker.go [hunk 4]
```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/spf13/afero"
 )
 
-var (
+const (
```
Comment (xh3b4sd) [inline workflow/docker.go]: Boyscouting I think having these immutable as constants is more right.

File: workflow/docker.go [hunk 5]
```diff
@@ -16,6 +16,14 @@ var (
...
 	DockerPushShaTaskName    = "docker-push-sha"
 	DockerPushLatestTaskName = "docker-push-latest"
+
+	// DockerImageRefFmt is the format string used to compute the reference of the
+	// Docker image used to build and push. It may look something like this.
+	//
+	//     quay.io/giantswarm/architect:master-e8363ac222255e991c126abe6673cd0f33934ac8
+	//
+	DockerImageRefFmt    = "%s/%s/%s:%s-%s"
+	LatestDockerImageTag = "latest"
```
Comment (xh3b4sd) [inline workflow/docker.go]: We use now constants and the format is documented and uses proper format placeholders. Please stop using `%v` for everything.

File: workflow/docker.go [hunk 6]
```diff
@@ -50,13 +58,7 @@ func NewDockerBuildTask(fs afero.Fs, projectInfo ProjectInfo) (tasks.Task, error
...
 			"build",
 			"-t",
-			fmt.Sprintf(
-				"%v/%v/%v:%v",
-				projectInfo.Registry,
-				projectInfo.Organisation,
-				projectInfo.Project,
-				projectInfo.Sha,
-			),
+			newDockerImageRef(projectInfo, projectInfo.Sha),
```
Comment (xh3b4sd) [inline workflow/docker.go]: A helper function makes this code a bit more DRY and clean.

File: workflow/docker.go [hunk 7]
```diff
@@ -72,14 +74,9 @@ func NewDockerRunVersionTask(fs afero.Fs, projectInfo ProjectInfo) (tasks.Task,
...
-				"%v/%v/%v:%v",
-				projectInfo.Registry,
-				projectInfo.Organisation,
-				projectInfo.Project,
-				projectInfo.Sha,
-			),
-			Args: []string{"version"},
+			Args:             []string{"version"},
+			Image:            newDockerImageRef(projectInfo, projectInfo.Sha),
+			WorkingDirectory: projectInfo.WorkingDirectory,
```
Comment (xh3b4sd) [inline workflow/docker.go]: I am not sure if I am breaking something now by fixing it. The working directory was empty before. The tests I wrote showed that. Should we be worried now?

File: workflow/docker_test.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
+package workflow
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/giantswarm/architect/tasks"
+)
+
+func Test_Workflow_Docker_NewDockerBuildTask(t *testing.T) {
```
Comment (xh3b4sd) [inline workflow/docker_test.go]: Here are actually some tests to make sure the args are right. LGTM.

File: workflow/docker_test.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func testNewProjectInfo() ProjectInfo {
+	return ProjectInfo{
+		WorkingDirectory: "/usr/code/",
+		Organisation:     "giantswarm",
+		Project:          "architect",
+
+		Branch: "master",
+		Sha:    "e8363ac222255e991c126abe6673cd0f33934ac8",
```
Comment (xh3b4sd) [inline workflow/docker_test.go]: I chose more real settings here to make the result more readable. It was super hard to parse for me using strings like `test-sha`. This data is more aligned with the real world.

File: workflow/workflow.go [hunk 10]
```diff
@@ -30,7 +30,9 @@ type ProjectInfo struct {
 	WorkingDirectory string
 	Organisation     string
 	Project          string
-	Sha              string
+
+	Branch string
+	Sha    string
```
Comment (xh3b4sd) [inline workflow/workflow.go]: Here is the branch actually added to the `ProjectInfo` type.

File: cmd/root.go (hunk 2 above)
Comment (rossf7) [inline cmd/root.go]: "within Giant Swarm release engineering" reads more easily for me.

File: workflow/docker.go (hunk 6 above)
Comment (rossf7) [inline workflow/docker.go]: <3

File: workflow/docker.go (hunk 7 above)
Comment (rossf7) [inline workflow/docker.go]: I don't know architect well enough to comment. @JosephSalisbury WDYT?

File: cmd/root.go (hunk 2 above)
Comment (xh3b4sd) [inline cmd/root.go]: Sweet. Done.

File: workflow/docker.go [hunk 11]
```diff
@@ -204,14 +179,20 @@ func NewDockerPushLatestTask(fs afero.Fs, projectInfo ProjectInfo) (tasks.Task,
...
-				projectInfo.Project,
-			),
+			newDockerImageRef(projectInfo, LatestDockerImageTag),
 		},
 	)
 
 	return dockerPush, nil
 }
+
+func newDockerImageRef(projectInfo ProjectInfo, version string) string {
```
Comment (JosephSalisbury) [inline workflow/docker.go]: I'd like to see a test over this.

File: workflow/docker.go (hunk 7 above)
Comment (JosephSalisbury) [inline workflow/docker.go]: I'm actually not certain here. If architect builds fine locally, then I don't see what else it could break.

**Comment by JosephSalisbury** [comment]
I would recommend trying to deploy, end to end, with the new tag scheme. "It should work" probably isn't good enough there, and we should make sure before merging this.

File: workflow/docker.go (hunk 11 above)
Comment (xh3b4sd) [inline workflow/docker.go]: I think this is already tested multiple times. Anyway added an explicit test.

File: workflow/docker.go (hunk 7 above)
Comment (xh3b4sd) [inline workflow/docker.go]: I could build architect locally with architect. Looks like this is fine.

**Comment by xh3b4sd** [comment]
> I would recommend trying to deploy, end to end, with the new tag scheme.

How to do this without merging it? I miss something here.

**Comment by JosephSalisbury** [comment]
Grab a helm chart for some service from quay, modify the version, push it, create a deployment event for that version.

**Comment by xh3b4sd** [comment]
I cannot test this locally. I cannot install helm and the registry plugin. The latest version is broken since a week. Don't know how to install another version. Need a pair on this.

**Comment by xh3b4sd** [comment]
For the record. I am removing the branch prefixing and keep the tests and stuff. See https://github.com/giantswarm/giantswarm/issues/1605#issuecomment-324326336.

**Comment by xh3b4sd** [comment]
Please have another look guys. I removed the branch prefix in the tags. I would like to keep the branch information within architect though. I do not want to waste time on cleaning this up and I think this information can be useful for logging, debugging, meta data at some point.
